powersimdata/design/generation/curtailment.py

```diff
@@ -0,0 +1,126 @@
+        either maintain given Pmin values (for dispatchable resources) or track profiles
+        (for profile resources, e.g. hydro).
+    :param dict/pandas.Series pmin_by_id: Mapping of IDs to Pmin assumptions, as an
+        override to the default behavior for that plant type. Values between 0 and 1
+        (inclusive) are treated as shares of Pmax, and None values either maintain given
+        Pmin values (for dispatchable resources) or track profiles
+        (for profile resources, e.g. hydro).
+    :param iterable curtailable: resource types which can be curtailed.
+    :return: (*float*) -- share of curtailable resources that will be curtailed.
+    :raises TypeError: if inputs do not mach specified type.
+    :raises ValueError: if any entries in curtailable or keys in pmin_by_type are not
+        types in the grid, any keys in pmin_by_id are not plant IDs in the Grid,
+        or any values in pmin_by_type/pmin_by_id are not in the range [0, 1] or None.
+    """
+    if not isinstance(scenario, Scenario):
+        raise TypeError("scenario must be a Scenario")
+    if pmin_by_type is None:
+        pmin_by_type = {}
+    if pmin_by_id is None:
+        pmin_by_id = {}
+    check_dicts = {"pmin_by_id": pmin_by_id, "pmin_by_type": pmin_by_type}
+    for name, d in check_dicts.items():
+        if not isinstance(d, (dict, pd.Series)):
+            raise TypeError(f"{name} must be a dict or pandas Series")
+        # Access values via appropriate method whether d is a dict or a pandas Series
+        values = d.values() if isinstance(d, dict) else d.values
+        if not all([v is None or 0 <= v <= 1 for v in values]):
+            err_msg = f"all entries in {name} must be None or in the range [0, 1]"
+            raise ValueError(err_msg)
+    plant = scenario.state.get_grid().plant
+    valid_types = plant["type"].unique()
+    if not set(pmin_by_type.keys()) <= set(valid_types):
+        raise ValueError("Got invalid plant type as a key to pmin_by_type")
+    if not set(pmin_by_id.keys()) <= set(plant.index):
+        raise ValueError("Got invalid plant id as a key to pmin_by_id")
+    try:
+        if not set(curtailable) <= set(valid_types):
+            raise ValueError("Got invalid plant type within curtailable")
+    except TypeError:
+        raise TypeError("curtailable must be an iterable")
+
+    # Get profiles, filter out plant-level overrides, then sum
+    all_profiles = pd.concat(
+        [getattr(scenario.state, m)() for m in set(profile_methods.values())], axis=1
+    )
+    plant_id_mask = ~plant.index.isin(pmin_by_id.keys())
+    base_plant_ids_by_type = plant.loc[plant_id_mask].groupby("type").groups
+    plant_ids_for_summed_profiles = set().union(
+        *[set(base_plant_ids_by_type[g]) for g in profile_methods]
+    )
+    summed_profiles = (
+        all_profiles[plant_ids_for_summed_profiles]
+        .groupby(plant.loc[plant_ids_for_summed_profiles, "type"], axis=1)
```

Comment:
I think we can simply do `all_profiles[plant_ids_for_summed_profiles].groupby(plant.type, axis=1).sum()` here.

Reply:
You're right. I thought there might be a problem since `plant.type` is a different length than `all_profiles[plant_ids_for_summed_profiles]`, but it seems to work since `plant.type` has an index where the appropriate entries can be looked up.

Reply:
Changed.